### PR TITLE
PLT-3633 EE: Add Google and Office365 SSO through OAuth2

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
* Adds Office 365 as an SSO provider through OAuth2 for E10
* Adds Google as an SSO provider through OAuth2 for E10
* Adds an OAuth2 tab to the System Console for E10

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3633

#### Checklist
- [x] Has enterprise changes (mattermost/enterprise#46)
- [x] Has UI changes
- [x] Includes text changes and localization files updated
- [x] Touches critical sections of the codebase (auth)

### CODE REVIEW

File: utils/config.go [hunk 1]
```diff
@@ -239,6 +239,7 @@ func getClientConfig(c *model.Config) map[string]string {
 
 	props["EnableSignUpWithGitLab"] = strconv.FormatBool(c.GitLabSettings.Enable)
 	props["EnableSignUpWithGoogle"] = strconv.FormatBool(c.GoogleSettings.Enable)
+	props["EnableSignUpWithOffice365"] = strconv.FormatBool(c.Office365Settings.Enable)
```
Comment (enahum) [inline utils/config.go]: I think this props (Google and Office365) should be set only if license like MFA, SAML, etc. don't you agree?

Comment (jwilander) [inline utils/config.go]: Yeah, not a big deal but it's nicer to do it that way. I'll update

**Comment by jasonblais** [comment]
Tested on Windows 10, Chrome

Proposed text updates submitted separately in https://github.com/mattermost/platform/pull/3677

**Google OAuth**
1) I got the error message below after enabling Google OAuth. I'm positive I entered the correct Client ID and Client Secret 
 - I tested this both with and without an existing account in the test server, corresponding to my Google accounts.

2) I think the error message should also read `Google OAuth not available on this server.` The `i18n/en.json` file uses a function {{.Service}} to determine which OAuth service provider it is (GitLab/Google/Office 365), so it wouldn't be a simple text change for me to do.

image

Tests left:
**Google OAuth**
1) Testing login and account creation successfully

**Office 365**
2) Is this something a PM can test easily..? Not sure I fully understand the tenant stuff..

3) Testing **System Console > Users and Teams > Restrict account creation to specified email domains** option works correctly.

**Comment by jwilander** [comment]
@jasonblais 

For Google attempted tests:
1) The spinmint PR build didn't include the new enterprise bits for some reason. It's fixed on the spinmint above. I wiped the database but signed up with the same credentials you gave me before.

2) I can update it to show the correct spelling and capitalization.

For Office 365 tests:
2) Yes, you can test by following the instructions. As long as you're signing in with the account you registered the app with then the testing should work no problem. When we expand the documentation onto our docs site I can go a bit more in-depth about the tenant stuff but there's no room to do it in the System Console directly.

3) I don't think this has ever worked with OAuth and I didn't update it to do so. I can add that functionality here but really it should be a different ticket

**Comment by jwilander** [comment]
Changes made, I will re-up the spinmint when the build finishes.

Only note is for the error message `google oauth not available on this server` it is now `Google SSO through OAuth 2.0 not available on this server` with the minor caveat that gitlab will be Gitlab and office365 will be Office365 instead of their proper grammatizations (I think I just invented that word). This is because it would have been annoying and mana-draining to do it with correct grammar and it's an error message that should very rarely be seen.

**Comment by jasonblais** [comment]
Thanks! 

Sounds good for the error message -- except just checking we're using `GitLab`, not `Gitlab` :)

**Comment by jwilander** [comment]
Nope it will be `Gitlab` and there will be no space in `Office365`. I could put mana into fixing it but I don't think it's worth the effort right now (it's more than it seems)

**Comment by jwilander** [comment]
@jasonblais spinmint updated. I had to wipe the DB again but I re-used the same credentials again and uploaded a license.

**Comment by jasonblais** [comment]
+1 from PM:

Tested on Windows, Chrome:

**Google OAuth**
 - account creation and login flow (permissions are asked only once, as expected)
 - switch between OAuth and email sign-in methods
 - testing various errors and corner cases
 - switching between different Google accounts

**Office 365**
 - tested set up, but final login was unsuccessful. Caused likely because spinmint uses `http` whereas Microsoft recommends `https` as redirect URI. Will do final testing with `https` URI once submitted to the CI.

One minor issue reported to JW, and I made a small tweak to help text. Otherwise all good. Docs will be submitted later.

**Comment by enahum** [comment]
+1

**Comment by crspeller** [comment]
+1

**Pull request merged**